libazureinit/src/status.rs

```diff
@@ -0,0 +1,385 @@
+}
+
+/// This function checks if the azure-init data directory is present, and if not,
+/// it creates it.
+fn check_provision_dir(config: Option<&Config>) -> Result<(), Error> {
+    let dir = get_provisioning_dir(config);
+    if !dir.exists() {
+        fs::create_dir_all(&dir)?;
+        tracing::info!("Created provisioning directory: {}", dir.display());
+    }
+    Ok(())
+}
+
+/// Determines if VM is a gen1 or gen2 based on EFI detection,
+/// Returns `true` if it is a Gen1 VM (i.e., not UEFI/Gen2).
+///
+/// # Parameters:
+/// - `mock_efi_path` (optional): Used in **tests** to override the default EFI detection path.
+///   - If provided, the function checks if the file exists:
+///     - If **the file exists**, it simulates Gen2 (`false`).
+///     - If **the file does not exist**, it simulates Gen1 (`true`).
+///   - If `None` is provided, it defaults to **checking real system paths** (`/sys/firmware/efi` and `/dev/efi`).
+fn is_vm_gen1(mock_efi_path: Option<&str>) -> bool {
+    if let Some(path) = mock_efi_path {
+        !Path::new(path).exists()
+    } else {
+        if Path::new("/sys/firmware/efi").exists()
+            || Path::new("/dev/efi").exists()
+        {
+            return false;
+        }
+        true
+    }
+}
+
+/// Converts the first three fields of a 16-byte array from big-endian to
+/// the native endianness, then returns it as a `Uuid`.
+///
+/// This partially swaps the UUID so that d1 (4 bytes), d2 (2 bytes), and d3 (2 bytes)
+/// are converted from big-endian to the local endianness, leaving the final 8 bytes as-is.
+fn swap_uuid_to_little_endian(mut bytes: [u8; 16]) -> Uuid {
+    let (d1, remainder) = bytes.split_at(std::mem::size_of::<u32>());
+    let d1 = d1
+        .try_into()
+        .map(u32::from_be_bytes)
+        .unwrap_or(0)
+        .to_ne_bytes();
+
+    let (d2, remainder) = remainder.split_at(std::mem::size_of::<u16>());
+    let d2 = d2
+        .try_into()
+        .map(u16::from_be_bytes)
+        .unwrap_or(0)
+        .to_ne_bytes();
+
+    let (d3, _) = remainder.split_at(std::mem::size_of::<u16>());
+    let d3 = d3
+        .try_into()
+        .map(u16::from_be_bytes)
+        .unwrap_or(0)
+        .to_ne_bytes();
+
+    let native_endian = d1.into_iter().chain(d2).chain(d3).collect::<Vec<_>>();
+    debug_assert_eq!(native_endian.len(), 8);
+    bytes[..native_endian.len()].copy_from_slice(&native_endian);
+    uuid::Uuid::from_bytes(bytes)
+}
+
+/// Retrieves the VM ID by reading `/sys/class/dmi/id/product_uuid` and byte-swapping if Gen1.
+///
+/// The VM ID is a unique system identifier that persists across reboots but changes
+/// when a VM is cloned or redeployed.
+///
+/// # Returns
+/// - `Some(String)` containing the VM ID if retrieval is successful.
+/// - `None` if something fails or the output is empty.
+pub fn get_vm_id() -> Option<String> {
+    private_get_vm_id(None, None)
+}
+
+fn private_get_vm_id(
+    custom_path: Option<&str>,
```

Comment:
ah this is better.  just a nit on name:

```suggestion
    product_uuid_path: Option<&str>,
    dev_efi_path: Option<&str>,
```